openeo/extra/job_management.py

```diff
@@ -121,17 +178,40 @@ def _normalize_df(self, df: pd.DataFrame) -> pd.DataFrame:
             df["geometry"] = df["geometry"].apply(shapely.wkt.loads)
         return df
 
-    # TODO: long method with deep nesting. Refactor it to make it more readable.
+    def _persists(self, df, output_file):
+        df.to_csv(output_file, index=False)
+        _log.info(f"Wrote job metadata to {output_file.absolute()}")
+
     def run_jobs(
         self, df: pd.DataFrame, start_job: Callable[[], BatchJob], output_file: Path
     ):
         """Runs jobs, specified in a dataframe, and tracks parameters.
 
         :param df:
             DataFrame that specifies the jobs, and tracks the jobs' statuses.
+
         :param start_job:
             A callback which will be invoked with the row of the dataframe for which a job should be started.
             This callable should return a :py:class:`openeo.rest.job.BatchJob` object.
+
+            In general the callback should have the following parameters described below,
+            If it does not need all of these you can use *args and **kwargs to ignore
+            the ones it doesn't need:
+
+                row:
+                    The row in the pandas dataframe that stores the jobs state and other tracked data.
+
+                connection_provider:
+                    Like connection in add_backend:
+                    - either a Connection to the backend,
+                    - or a callable to create a backend connection.
+
+                connection:
+                    The Connection itself, that has already been created.
+
+                provider:
+                    The name of the backend that will run the job.
+
```

Comment:
1) The way we define what start_job should look like is a bit awkward. I think start_job would be a good candidate for using a Protocol, that defines and documents what the expected / supported function's parameters are, but Protocol was only introduced in Python 3.8.
Maybe a TODO for when we drop support for python 3.7 and 3.6?

2) That said we need to document what parameters the job manager passes to start_jobs so we know what parameters you can (maximally) have in your implementation.
This way to document it is a little clunky but I think enough for now.
Suggestions for a better solution are welcome.

3) We allow two similar parameters which is a little confusion: both a Connection as well as a callback to create a connection.
But I think the reason was that in practice the implementation would use either one or the other, if it is using them at all. If this is correct, that can help improve to improve the docstring.

Also, it looks like the intention is that start_job always uses keyword arguments and `*args, **kwargs`, so you can simply leave out parameters that you aren't going to use, even though the job manager will pass them to start_jobs.

Reply:
I updated the docstring to make it more clear about the points above.
That is probably good enough for now.